numpy/lib/function_base.py

```diff
@@ -91,16 +102,37 @@ def _hist_optim_numbins_estimator(a, estimator):
     if a.size == 0:
         return 1
 
-    def sturges(x):
+    # Filter the data so we stay in the required range
+    if data_range is not None:
+        mn, mx = data_range
+        keep = (a >= mn)
+        keep &= (a <= mx)
+        if not np.logical_and.reduce(keep):
```

Comment:
I coped the filter step from https://github.com/numpy/numpy/blob/master/numpy/lib/function_base.py#L400 which came up in #6100. 

It's actually faster than `np.all`